### PR TITLE
Minor improvements

### PR DESCRIPTION
**Pull request by MarijnS95**
Many minor improvements.

- Automatic patch discovery;
- Automatic packet+packetprocessor discovery (+ unittests);
- Refactor of player functions into class;
- Player rotation (and animationrotation sync);
- Live reloading of methods (TODO: write a little wiki page detailing this);
- And some other changes.

### CODE REVIEW

File: NitroxPatcher/Main.cs [hunk 1]
```diff
@@ -1,49 +1,45 @@
...
             HarmonyInstance.DEBUG = false;
 
             HarmonyInstance harmony = HarmonyInstance.Create("com.nitroxmod.harmony");
 
-            foreach(NitroxPatch patch in patches)
-            {
-                Console.WriteLine("[NITROX] Applying " + patch.GetType());
-                patch.Patch(harmony);
-            }
+            //AppDomain.CurrentDomain.GetAssemblies()
```
Comment (Enflorence) [inline NitroxPatcher/Main.cs]: Minor: remove commented code

File: NitroxReloader/Reloader.cs [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+                var handle = GetRuntimeMethodHandle(method);
+                RuntimeHelpers.PrepareMethod(handle);
+                return handle.GetFunctionPointer();
+            }
+
+            public static unsafe void WriteJump(IntPtr original, IntPtr destination)
+            {
+                Console.WriteLine("Reloader: Writing jump to {0} at {1}", destination, original);
+                bool x64 = IntPtr.Size == sizeof(long);
+#if true
```
Comment (Enflorence) [inline NitroxReloader/Reloader.cs]: Is this directive conditional necessary?

File: NitroxReloader/Reloader.cs [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+    public class Reloader
+    {
+        private static readonly HashSet<string> assemblyWhitelist = new HashSet<string>() {
+            "NitroxModel.dll",
+            "NitroxClient.dll",
+            "NitroxPatcher.dll",
+        };
+
+        // TODO: Reloadable attribute for classes? (To define all methods in there to be reloadable)
```
Comment (Enflorence) [inline NitroxReloader/Reloader.cs]: It might be better to define these TODOs as tasks in Trello than in code.

File: NitroxReloader/Reloader.cs [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+// Modified to remove RimWorld dependency, and cleanups.
+
+namespace NitroxReloader
+{
+    [AttributeUsage(AttributeTargets.Method)]
+    public class ReloadableMethod : Attribute
+    {
+    }
+
+    public class Reloader
```
Comment (Enflorence) [inline NitroxReloader/Reloader.cs]: I would consider refactoring the classes into individual files to make it easier to maintain and evolve.

File: NitroxReloader/Reloader.cs (hunk 2 above)
Comment (MarijnS95) [inline NitroxReloader/Reloader.cs]: It was to quickly toggle between the two methods of writing the jump, but I couldn't figure out why the second one doesn't do what it's supposed to do so I removed it.

File: NitroxClient/GameLogic/PlayerManager.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            set
+            {
+                if (value == null)
+                    RemovePlayer(playerId);
+                else
+                    playersById[playerId] = value;
+            }
+        }
+
+        public RemotePlayer FindPlayer(string playerId, bool createPlayer = false)
```
Comment (Sunrunner37) [inline NitroxClient/GameLogic/PlayerManager.cs]: Optional RemotePlayer to encapsulate null OR always instantiate it if it is not found.  I don't think there is really a good reason for a null reference to be floating around.

File: NitroxClient/GameLogic/PlayerManager.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            RemotePlayer player;
+            if (!playersById.TryGetValue(playerId, out player) && createPlayer)
+                player = playersById[playerId] = new RemotePlayer(playerId);
+            return player;
+        }
+
+        public void ForPlayer(string playerId, Action<RemotePlayer> action, bool createPlayer = false)
+        {
+            var player = FindPlayer(playerId, createPlayer);
+            if (player != null)
```
Comment (Sunrunner37) [inline NitroxClient/GameLogic/PlayerManager.cs]: Please always use curly braces - no one liners like this.  There are a few other places in the PR - do a quick glance through.

File: NitroxClient/GameLogic/PlayerManager.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+        public RemotePlayer FindPlayer(string playerId, bool createPlayer = false)
+        {
+            RemotePlayer player;
+            if (!playersById.TryGetValue(playerId, out player) && createPlayer)
+                player = playersById[playerId] = new RemotePlayer(playerId);
+            return player;
+        }
+
+        public void ForPlayer(string playerId, Action<RemotePlayer> action, bool createPlayer = false)
```
Comment (Sunrunner37) [inline NitroxClient/GameLogic/PlayerManager.cs]: Can you make this method name more descriptive.  It isn't immediately obvious what it does

File: NitroxClient/GameLogic/PlayerManager.cs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        {
+            var player = FindPlayer(playerId, createPlayer);
+            if (player != null)
+                action(player);
+        }
+
+        public void RemovePlayer(string playerId)
+        {
+            playersById[playerId].Destroy();
```
Comment (Sunrunner37) [inline NitroxClient/GameLogic/PlayerManager.cs]: Looks like a possible NRE if the player does not exist. (This is the type of situation that optional protects against)

File: NitroxModel/Packets/VehicleMovement.cs [hunk 9]
```diff
@@ -21,7 +19,7 @@ public VehicleMovement(String playerId, Vector3 playerPosition, Quaternion rotat
 
         public override string ToString()
         {
-            return "[VehicleMovement - playerId: " + PlayerId + " position: " + PlayerPosition + " Rotation: " + Rotation + " TechType: " + TechType + " guid: " + Guid + "]";
+            return "[VehicleMovement - TechType: " + TechType + " guid: " + Guid + "]\n\t" + base.ToString();
```
Comment (Sunrunner37) [inline NitroxModel/Packets/VehicleMovement.cs]: base.toString()?

Comment (MarijnS95) [inline NitroxModel/Packets/VehicleMovement.cs]: Yes, the base class of `VehicleMovement` (which is `Movement`) already stringifies these fields, so it's better to reuse that functionality than to copy it.

**Pull request merged**